### PR TITLE
Parallel module install

### PR DESCRIPTION
**Pull request by lizmat**
An attempt at parallelizing module installation, now that a blocking race condition has been fixed by nine.

### CODE REVIEW

File: src/core.c/CompUnit/Repository/Installation.pm6 [hunk 1]
```diff
@@ -148,147 +148,200 @@ sub MAIN(:$name, :$auth, :$ver, *@, *%) {
...
+                        note "(Already doing $id)" if $verbose;
+                        next;
+                    }
+
+                    my $source      := $sources-dir.add($id);
+                    my $source-file := $repo-prefix
+                      ?? $repo-prefix ~ $source.relative($.prefix)
+                      !! $source;
+
+                    @promises.push: start {
```
Comment (ugexe) [inline src/core.c/CompUnit/Repository/Installation.pm6:321]: Spawning a process for every module listed under provides is going to reach diminishing returns pretty quickly.

**Comment by niner** [comment]
> Spawning a process for every module listed under provides is going to reach
> diminishing returns pretty quickly.

The problem is that we don't know the dependencies between the modules in a 
dist. Ordinarily you'd walk the tree bottom up and keep a couple processes 
busy that way, but we don't know how the nodes connect to a tree. Doesn't help 
if you spawn X processes but they all are waiting for the same dependency, 
while there may be other leaf modules available.

So the next step seems to be a large and hard one involving those precomp 
processes communicating their status (compiling/waiting for another precomp) 
to the main process so the latter can throttle accordingly. And even then you 
need even more communication as finishing one dependency may suddenly unblock 
a whole horde of precomp processes.

**Comment by vrurg** [comment]
I did not follow the whole story of parallelizing precomps and may miss a thing. But how hard would it be to make it a task queue based processing? So that limited number of worker threads are pre-spawned and each would pick a task from a channel?

**Comment by ugexe** [comment]
> I did not follow the whole story of parallelizing precomps and may miss a thing. But how hard would it be to make it a task queue based processing? So that limited number of worker threads are pre-spawned and each would pick a task from a channel?

Precompilation currently spawns a new process and thus requires IPC

**Comment by lizmat** [comment]
Closing this now, as it appears we should be able to solve slow installation better with making sure we only precomp a module once in the installation process.

**Pull request closed without merging**